[FLINK-25708] Bump Flink dependency to 1.14.3

tillrohrmann opened the pull request:
This PR bumps the Flink dependency to 1.14.3. I had to add some dependency management entries to resolve conflicting transitive dependencies.

Optional: This PR also contains a very simple migration test based on the `EmbeddedSmokeHarnessTest`. This test is not ignored as the other `Harness` tests are.

igalshilman commented on statefun-flink/statefun-flink-harness/src/main/java/org/apache/flink/statefun/flink/harness/Harness.java:
```diff
@@ -126,6 +129,14 @@ public Harness withSavepointLocation(String savepointLocation) {
+    }
+  }
+
+  public JobClient startAsync() throws Exception {
```
I'm not sure that I'd like to expose the `JobClient` as a public method on the `Harness`.
How about we think about it as a followup PR?

tillrohrmann replied:
Makes sense.

igalshilman commented on statefun-e2e-tests/statefun-smoke-e2e-embedded/src/test/java/org/apache/flink/statefun/e2e/smoke/embedded/EmbeddedSmokeHarnessMigrationTest.java:
```diff
@@ -0,0 +1,127 @@
+import org.slf4j.LoggerFactory;
+
+@RunWith(Parameterized.class)
+public class EmbeddedSmokeHarnessMigrationTest {
```
Thanks for adding a migration test!
This test would catch any internal migration issues with the core statefun job graph.

Perhaps we need to followup with a separate e2e migration test, that runs within its own migration profile, that includes at least a Kafka ingress/egress pair.
It can be completely based remote functions, without any verification/failure injection aspects to it, and this probably should be based on `testcontainers` instead of the harness.

having test containers will allows the test to create a save point in an arbitrary previous version (say 3.1.1) 
based on the previously released official docker images, and load it with the current version.
The test would be bit more time consuming, and that's why I would recommend that to run under a different profile.

If this is something you find reasonable, then I'll create a followup JIRA issue.

tillrohrmann replied:
I think this sounds like a good idea. Having these test to ensure that nothing is breaking unintentionally is a good improvement. I will remove the commit that introduced the harness test and we continue based on the migration test JIRA ticket you'll create.

igalshilman reviewed (approved):
Thanks @tillrohrmann for migrating StateFun for the latest Flink version!
I've run the end to end tests locally, and they pass.

I have one comment about exposing Flink`s `JobClinet` in the `Harness` , for your consideration, and a followup regarding a migration test.

Thanks again!

tillrohrmann commented:
Thanks a lot for your review @igalshilman. I've removed the migration test commit and rebased onto the latest master. Will merge it once GHA gives green light.

I'd be +1 for following up with a dedicated JIRA ticket for adding a migration test. It will help spotting incompatible changes faster.

cosmir17 commented:
The pr has been merged 🎉🥳 I should just wait. Any news on public dependency release? @tillrohrmann 😅

tillrohrmann commented:
I will propose to create a new release today. So expect that the release happens later this week.

cosmir17 commented:
Thank you for your kindness 🥳🥰 @tillrohrmann

tillrohrmann commented:
@cosmir17 if you want to help with the release verification, then I can ping you once we have the release artifacts created.
